### ISSUE TITLE
Custom dock and undock command Marlin

### ISSUE THREAD

**Issue by chris-hemmings** [MarlinFirmware/Marlin#12646]
I have custom G codes i have written into marlin 1.1.0 and i have only just been able to test it properly with another printer. 

Im waiting on my friend to upload 1.1.9 to his printer to test it. But someone might know a reason for the pause.

I have a few issues and I was wondering if someone could help me out.

following on from post #5770
https://github.com/MarlinFirmware/Marlin/issues/5770

After the head is docked or undocked, there is pause, before anything continues.

See video https://youtu.be/-WE_fNBL_Z0

This is the finished code in configuration.h

```
//===========================================================================
//========================= Tool Dock and undock ============================
//===========================================================================
#define TOOL_NUM        2 // number of tools

#define DOCKN_PIN       {X_MAX_PIN, Y_MIN_PIN} //switches assigned to these pins

#define MOVES_IN_DOCK_SEQUENCE  4 //number of blocking moves in tool docking procedure
      
#define DOCK_X                  {{70, 70, 8, 8},     /*Dock T0 X coords*/    {162, 162, 224, 224}    /*Dock T1 X coords*/    }    //So docking T0 will move to X70 Y10, X70 Y-31, X11 Y-31 X8 Y10, 
#define DOCK_Y                  {{50, 0, 0, 50}, /*Dock T0 Y coords*/    {50, 0, 0, 50}  /*Dock T1 Y coords*/    }    // docking T1 will move to X162 Y10, X162 Y-31, X222 Y-31, X222 Y10
#define DOCK_FR                 {{10000, 10000, 10000, 10000}, /*Dock T0 Feedrates*/    {10000, 10000, 10000, 10000}    /*Dock T1 Feedrates*/    }   
#define Z_LIFT_HEIGHT           10
#define Z_LIFT_SPEED            1200 //Z lift/lower speed in mm/min 

#define E_PRIME                 20 //how much to prime Extruder
#define E_PRIME_FEEDRATE        300 //Feedrate for prime
```

and in Marlin_main.cpp i have this 

```
inline void tool_dock_undock(bool undock) {
  //if undock is set to false. we are docking. otherwise we are undocking
  #define IS_EMPTY(PIN)   (!digitalRead(PIN))
  
  int tool = -1;
  int dock = -1;
  
  

  int dockPin[] = DOCKN_PIN;
  float dockX[][MOVES_IN_DOCK_SEQUENCE] = DOCK_X;
  float dockY[][MOVES_IN_DOCK_SEQUENCE] = DOCK_Y; 
  float dockFR[][MOVES_IN_DOCK_SEQUENCE] = DOCK_FR; 

  if (! (axis_known_position[X_AXIS] && axis_known_position[Y_AXIS]) )
  {
   SERIAL_ECHO_START;
    SERIAL_ECHOLN("Please home before calling");
    return; // abort, since we don't know where we are
  }
  
  if (code_seen('T')) {
     tool = code_value_int();
     
     
  } else {
    //if no tool is specified we send a message and exit here
    SERIAL_ECHOLN("Please specify a tool using the T parameter.");
    return;
  }
  //check that the selected tool is valid (i.e T5 is invalid if we only have two tools)
  if(tool < 0 || tool >= TOOL_NUM) {
    //warn and exit if invalid
    SERIAL_ECHO("Warning! Please pick a tool from 0 to ");
    SERIAL_ECHOLN(TOOL_NUM - 1);
    return;
  }

  



  //check if the dock is empty (if docking) or full (if undocking)
  if(undock ? !IS_EMPTY(dockPin[tool]) : IS_EMPTY(dockPin[tool])) {
    dock = tool;
  } else {
    //that tool is already docked, error and exit
    if(!undock) {
      SERIAL_ECHOLN("Tool is currently in that dock!");
    } else {
      SERIAL_ECHOLN("No tool detected in that dock!");
    }
    return;
  }

  //if dock is empty, dock there
  if(dock != -1) {

    //"Docking T0..."
    if(undock) {
      SERIAL_ECHO("Undocking ");
    } else {
      SERIAL_ECHO("Docking ");
    }
    SERIAL_ECHO("T");
    SERIAL_ECHO(tool);
    SERIAL_ECHOLN("...");
    
    
    //Lift z by 10mm or Drop z by 10mm for beds that move
    do_blocking_move_to_z(current_position[Z_AXIS]+Z_LIFT_HEIGHT, Z_LIFT_SPEED);

    
    //do blocking moves to dock

    int loopStart = (undock ? (MOVES_IN_DOCK_SEQUENCE-1) : 0);
    int loopEnd   = (undock ? 1 : MOVES_IN_DOCK_SEQUENCE-2);
    int loopInc   = (undock ? -1 : 1);
    
    for(int i = loopStart; i != loopEnd; i+=loopInc) {                  //Perform Movements in X,Y and Feedrates
      do_blocking_move_to_xy(dockX[dock][i], dockY[dock][i], dockFR[dock][i]);      
    }  

if(undock) {
  tool_change(tool);
       set_destination_to_current();
            planner.buffer_line(
                 current_position[X_AXIS], current_position[Y_AXIS], current_position[Z_AXIS],
                 current_position[E_AXIS]+= E_PRIME, MMM_TO_MMS(E_PRIME_FEEDRATE), active_extruder );
           stepper.synchronize();
           set_destination_to_current();
}
      loopStart = (undock ? (MOVES_IN_DOCK_SEQUENCE-3) : 2);
      loopEnd   = (undock ? -1 : MOVES_IN_DOCK_SEQUENCE);
      loopInc   = (undock ? -1 : 1);
    
    for(int i = loopStart; i != loopEnd; i+=loopInc) {                  //Perform Movements in X,Y and Feedrates
      do_blocking_move_to_xy(dockX[dock][i], dockY[dock][i], dockFR[dock][i]);      
    }   

                                                                     
    //Drop z by 10mm or Lift z by 10mm for beds that move                                                              
   do_blocking_move_to_z(current_position[Z_AXIS]-Z_LIFT_HEIGHT, Z_LIFT_SPEED);

    if(undock) {
      SERIAL_ECHO("Undocking done.");
    } else {
      SERIAL_ECHO("Docking done.");
    }
  }
}

/**
 * End Dock Command
 */
```

and 

```
case 7: // G7 Dock Tool 
        tool_dock_undock(false);
        break;

      case 8: // G8 Undock Tool
        tool_dock_undock(true);
        break;
```

If I do a G7 command following by a G6 Command (dock one tool and pick up the other), using a gcode there is no pause. Only when printing a file.

Any ideas?

**Comment by wookie666** [comment]
[https://youtu.be/-WE_fNBL_Z0](url)

**Comment by chris-hemmings** [comment]
Nevermind, Fixed by updating to 1.1.9 bugfix and altering my code
